File: typescript/src/TennisGame3.ts
import { TennisGame } from './TennisGame';


export class TennisGame3 implements TennisGame {
  private p2: number = 0;
  private p1: number = 0;
  private p1N: string;
  private p2N: string;

  constructor(p1N: string, p2N: string) {
    this.p1N = p1N;
    this.p2N = p2N;
  }

  getScore(): string {
    let s: string;
    if (this.p1 < 4 && this.p2 < 4 && !(this.p1 + this.p2 === 6)) {
      const p: string[] = ['Love', 'Fifteen', 'Thirty', 'Forty'];
      s = p[this.p1];
      return (this.p1 === this.p2) ? s + '-All' : s + '-' + p[this.p2];
    } else {
      if (this.p1 === this.p2)
        return 'Deuce';
      s = this.p1 > this.p2 ? this.p1N : this.p2N;
      return (((this.p1 - this.p2) * (this.p1 - this.p2)) === 1) ? 'Advantage ' + s : 'Win for ' + s;
    }
  }

  wonPoint(playerName: string): void {
    if (playerName === 'player1')
      this.p1 += 1;
    else
      this.p2 += 1;

  }
}

File: dart/lib/TennisGame3.dart

import 'dart:core';

import 'TennisGame.dart';

class TennisGame3 implements TennisGame {

    int _p2 = 0;
    int _p1 = 0;
    String _p1N;
    String _p2N;

    TennisGame3(String p1N, String p2N) {
        this._p1N = p1N;
        this._p2N = p2N;
    }

    String getScore() {
        String s;
        if (_p1 < 4 && _p2 < 4 && !(_p1 + _p2 == 6)) {
            List<String> p = ["Love", "Fifteen", "Thirty", "Forty"];
            s = p[_p1];
            return (_p1 == _p2) ? s + "-All" : s + "-" + p[_p2];
        } else {
            if (_p1 == _p2)
                return "Deuce";
            s = _p1 > _p2 ? _p1N : _p2N;
            return ((_p1-_p2)*(_p1-_p2) == 1) ? "Advantage " + s : "Win for " + s;
        }
    }

    void wonPoint(String playerName) {
        if (playerName == "player1")
            this._p1 += 1;
        else
            this._p2 += 1;

    }

}
